warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java

```diff
@@ -0,0 +1,431 @@
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.ASTVar;
+import org.apache.commons.jexl2.parser.ASTWhileStatement;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.SimpleNode;
+import org.apache.log4j.Logger;
+
+/**
+ * Detect if a query tree contains nodes surrounded by multiple reference expressions, or a child node with an extra wrap.
+ *
+ * Assumes that each ASTReferenceExpression has a parent ASTReference node
+ */
+public class MinimalReferenceExpressionsVisitor extends BaseVisitor {
+    
+    private static final Logger log = Logger.getLogger(MinimalReferenceExpressionsVisitor.class);
+    
+    private boolean isValid = true;
+    
+    public static boolean validate(JexlNode node) {
+        MinimalReferenceExpressionsVisitor visitor = new MinimalReferenceExpressionsVisitor();
+        node.jjtAccept(visitor, null);
+        return visitor.isValid();
+    }
+    
+    public boolean isValid() {
+        return this.isValid;
+    }
+    
+    /*
+     * Operations
+     */
+    
+    @Override
+    public Object visit(ASTReference node, Object data) {
+        
+        if (isValid && isParen(node) && isParen(grandChild(node))) {
+            // found a double paren
+            invalidate(node, "double paren");
+            return data;
+        } else if (!isValid) {
+            return data;
+        }
+        
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        if (isValid && isAnyChildAWrappedSingleTerm(node)) {
+            // found a single term child that is wrapped
+            invalidate(node, "wrapped single child");
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        if (!isValid)
+            return data;
+        
+        QueryPropertyMarker.Instance instance = QueryPropertyMarker.findInstance(node);
+        if (instance.isAnyType()) {
+            // descend into the marker's source node
+            JexlNode source = instance.getSource();
+            source.childrenAccept(this, data);
+            return data;
+        } else if (isAnyChildAWrappedSingleTerm(node)) {
+            // invalidate
+            invalidate(node, "wrapped single child");
+            return data;
+        } else {
+            // otherwise continue descent
+            node.childrenAccept(this, data);
+            return data;
+        }
+    }
+    
+    /*
+     * Helper methods
+     */
+    
+    // is this node a parentheses
+    private boolean isParen(JexlNode node) {
+        return node instanceof ASTReference && node.jjtGetNumChildren() == 1 && node.jjtGetChild(0) instanceof ASTReferenceExpression;
+    }
+    
+    // fetch the grandchild
+    private JexlNode grandChild(JexlNode node) {
```

Comment:
Might be worth clarifying `getFirstGrandchild` or `getLeftmostGrandchild` or something? Just to specify that you are targeting a specific one out of possibly many.

Reply:
Worth doing, thanks for pointing this out!